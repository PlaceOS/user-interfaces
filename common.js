"use strict";(self.webpackChunkworkplace=self.webpackChunkworkplace||[]).push([["common"],{7625:(W,T,t)=>{t.d(T,{I:()=>_});var L=t(3918),g=t(9638),D=t(8368),M=t(1299),I=t(6221),B=t(2886),U=t(3690),p=t(1810),O=t(9377),c=t(3200),u=t(4505),y=t(7716),C=t(823),o=t(8759),x=t(9151),P=t(6116),v=t(9095),l=t(3298),h=t(6942),a=t(9128),r=t(7418),K=t(9676);class _ extends D.KG{constructor(i,d){super(),this._settings=i,this._org=d,this._poll=new u.X(0),this._poll_type=new u.X("api"),this._loading=new u.X(!1),this._filters=new u.X({shown_types:["event","desk","parking","visitor"]}),this._date=new u.X(Date.now()),this._update=(0,y.aj)([this._date,this._poll]).pipe((0,C.b)(500),(0,o.b)(s=>this._loading.next(!0))),this._space_bookings=this._org.active_building.pipe((0,x.h)(s=>!!s),(0,P.g)("id"),(0,C.b)(300),(0,o.b)(s=>this.unsubWith("bind:")),(0,v.w)(({id:s})=>(this._loading.next(!0),(0,B.u2)(s))),(0,l.x)(([s],[e])=>s!==e),(0,v.w)(s=>(this._loading.next(!1),(0,y.aj)((s||[]).map(e=>{const n=(0,U.rTZ)(e.id,"Bookings").binding("bookings"),f=n.listen().pipe((0,h.U)(R=>(R||[]).map(A=>new M.ym({...A,resources:A.attendees.filter(m=>m.email===e.email||m.resource),system:e}))));return this.hasSubscription(`bind:${e.id}`)||this.subscription(`bind:${e.id}`,n.bind()),f})))),(0,h.U)(s=>(0,D.xH)(s)),(0,a.d)(1)),this.ws_events=this._space_bookings.pipe((0,h.U)(s=>{const e=(0,D.ar)();return s.filter(n=>n.host.toLowerCase()===e.email.toLowerCase()||n.attendees.find(f=>f.email.toLowerCase()===e.email.toLowerCase()))})),this.api_events=this._update.pipe((0,v.w)(([s])=>{const e={period_start:(0,p.Z)((0,O.Z)(s)),period_end:(0,p.Z)((0,c.Z)(s))};return this._settings.get("app.events.use_bookings")?(0,g.F2)({...e,type:"room"}).pipe((0,h.U)(n=>n.map(f=>(0,M.Yd)(f))),(0,r.K)(n=>[])):(0,M.u$)({...e}).pipe((0,r.K)(n=>[]))}),(0,o.b)(()=>this.timeout("end_loading",()=>this._loading.next(!1))),(0,a.d)(1)),this.events=this._poll_type.pipe((0,v.w)(s=>"api"===s?this.api_events:this.ws_events)),this.visitors=this._update.pipe((0,v.w)(([s])=>(0,g.F2)({period_start:(0,p.Z)((0,O.Z)(s)),period_end:(0,p.Z)((0,c.Z)(s)),type:"visitor"}).pipe((0,r.K)(e=>(console.error(e),[])))),(0,o.b)(()=>this.timeout("end_loading",()=>this._loading.next(!1))),(0,a.d)(1)),this.desks=this._update.pipe((0,v.w)(([s])=>(0,g.F2)({period_start:(0,p.Z)((0,O.Z)(s)),period_end:(0,p.Z)((0,c.Z)(s)),type:"desk"}).pipe((0,r.K)(e=>(console.error(e),[])))),(0,o.b)(()=>this.timeout("end_loading",()=>this._loading.next(!1))),(0,a.d)(1)),this.parking=this._update.pipe((0,v.w)(([s])=>(0,g.F2)({period_start:(0,p.Z)((0,O.Z)(s)),period_end:(0,p.Z)((0,c.Z)(s)),type:"parking"}).pipe((0,r.K)(e=>[]))),(0,o.b)(()=>this.timeout("end_loading",()=>this._loading.next(!1))),(0,a.d)(1)),this.bookings=(0,y.aj)([this.events,this.visitors,this.desks,this.parking]).pipe((0,h.U)(([s,e,n,f])=>[...s,...e,...n,...f].sort((R,A)=>R.date-A.date))),this.filtered_bookings=(0,y.aj)([this.bookings,this._filters]).pipe((0,h.U)(([s,e])=>s.filter(n=>n instanceof M.ym&&e?.shown_types?.includes("event")||e?.shown_types?.includes(n.booking_type)))),this.filters=this._filters.asObservable(),this.date=this._date.asObservable(),this.loading=this._loading.asObservable(),this.subscription("poll_type",this._org.active_building.subscribe(()=>this._poll_type.next(this._settings.get("app.schedule.use_websocket")?"ws":"api")))}triggerPoll(){this._poll.next(Date.now())}startPolling(i=6e4){return this.interval("poll",()=>{"visible"===document.visibilityState&&this._poll.next(Date.now())},i),()=>this.stopPolling()}stopPolling(){this.clearInterval("poll")}setDate(i){this._date.next(i)}removeItem(i){this._poll.next(Date.now())}toggleType(i,d=!1){var s=this;return(0,L.Z)(function*(){const e=s._filters.getValue()||{shown_types:[]},{shown_types:n}=e;n&&(n.includes(i)||d)?s._filters.next({...e,shown_types:n.filter(f=>f!==i)}):s._filters.next({...e,shown_types:[...n,i]})})()}}_.\u0275fac=function(i){return new(i||_)(K.LFG(D.gb),K.LFG(I.w7))},_.\u0275prov=K.Yz7({token:_,factory:_.\u0275fac,providedIn:"root"})},8676:(W,T,t)=>{t.d(T,{I:()=>r});var L=t(9638),g=t(9434),D=t(8368),M=t(1299),I=t(1810),B=t(9377),U=t(312),p=t(3200),O=t(4505),c=t(1942),u=t(7716),y=t(4350),C=t(9095),o=t(9128),x=t(823),P=t(522),v=t(6942),l=t(7418),h=t(8759),a=t(9676);class r extends D.KG{constructor(_){super(),this._settings=_,this._poll=new O.X(0),this._options=new O.X({start:Date.now()}),this._loading=new O.X(""),this._schedule=new O.X([]),this.options=this._options.asObservable(),this.loading=this._loading.asObservable(),this.schedule=this._loading.asObservable(),this.calendars=(0,c.H)(1e3).pipe((0,C.w)(E=>(0,g.yJ)()),(0,o.d)(1)),this.events=(0,u.aj)([this._options,this._poll]).pipe((0,x.b)(1e3),(0,P.zg)(([E])=>{this._loading.next("Loading schedule...");const i={period_start:(0,I.Z)((0,B.Z)(E.start)),period_end:(0,I.Z)((0,U.Z)((0,p.Z)(E.start),6))};return E.calendar&&(i.calendar=E.calendar),this._schedule.next(this._schedule.getValue().filter(d=>!(0,D.MZ)(1e3*i.period_start,1e3*i.period_end,d.date,d.date+60*d.duration*1e3))),(0,y.D)([!0===this._settings.get("app.events.use_bookings")?(0,L.F2)({...i,type:"room"}).pipe((0,v.U)(d=>d.map(s=>(0,M.Yd)(s)))):(0,M.u$)({...i}),(0,L.F2)({...i,type:"desk"}),(0,L.F2)({...i,type:"parking"})]).pipe((0,l.K)(d=>[]))}),(0,v.U)(([E,i])=>{const d=[...this._schedule.getValue(),...E,...i.filter(s=>"declined"!==s.status)].sort((s,e)=>s.date-e.date);return this._schedule.next((0,D.Tw)(d,"id")),d}),(0,l.K)(E=>[]),(0,h.b)(E=>this._loading.next("")),(0,o.d)(1))}startPolling(_=15e3){this.interval("poll",()=>this._poll.next(Date.now()),_)}stopPolling(){this.clearInterval("poll")}setOptions(_){this._options.next({...this._options.getValue(),..._})}}r.\u0275fac=function(_){return new(_||r)(a.LFG(D.gb))},r.\u0275prov=a.Yz7({token:r,factory:r.\u0275fac,providedIn:"root"})},9434:(W,T,t)=>{t.d(T,{ot:()=>P,yJ:()=>o.yJ}),t(4603);var g=t(3918),D=t(4505),M=t(8759),I=t(9128),B=t(5670),U=t(1810),p=t(9377),O=t(3200),c=t(2317),u=t(8939),y=t(719),C=t(565),o=t(6217),x=t(9676);class P extends u.K{constructor(l,h){super(),this._org=l,this._settings=h,this._calendars=new D.X([]),this.calendar_list=(0,o.yJ)().pipe((0,M.b)(a=>this._calendars.next(a)),(0,I.d)(1)),this.query=()=>(0,o.yJ)(),this.freeBusy=a=>(0,o.tS)(a,this._org),this.availability=a=>(0,o.v7)(a),this._org.initialised.pipe((0,B.P)(a=>a)).subscribe(()=>this.init())}init(){var l=this;return(0,g.Z)(function*(){l._settings.get("app.events.use_bookings")||l._initialised.next(!0)})()}get calendars(){return this._calendars.getValue()}getFreeBusyDate(l,h){return(0,o.tS)({period_start:(0,U.Z)((0,p.Z)(l)),period_end:(0,U.Z)((0,O.Z)(l)),calendars:h},this._org)}checkSpacesAvailability(l,h,a,r){return(0,g.Z)(function*(){const K=yield(0,o.v7)({period_start:h,period_end:a,system_ids:l.join(",")}).toPromise(),_=new Date(r?.date).valueOf(),E=(0,c.Z)(_,r?.duration).valueOf();return!!K.every(d=>{const s=d.availability;if(r&&d.id===r.system?.email){const e=s.findIndex(n=>n.date>=_&&(0,c.Z)(n.date,n.duration).valueOf()<=E);-1!==e&&s.splice(e,1)}return!s.length})})()}}P.\u0275fac=function(l){return new(l||P)(x.LFG(C.w),x.LFG(y.g))},P.\u0275prov=x.Yz7({token:P,factory:P.\u0275fac,providedIn:"root"})}}]);
//# sourceMappingURL=common.js.map