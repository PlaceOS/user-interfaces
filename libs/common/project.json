{
    "name": "common",
    "$schema": "../../node_modules/nx/schemas/project-schema.json",
    "projectType": "library",
    "sourceRoot": "libs/common/src",
    "prefix": "placeos",
    "targets": {
        "lint": {
            "executor": "@nx/eslint:lint",
            "outputs": ["{options.outputFile}"]
        },
        "test": {
            "executor": "@nx/jest:jest",
            "outputs": ["{workspaceRoot}/coverage/libs/common"],
            "options": {
                "jestConfig": "libs/common/jest.config.ts"
            }
        }
    },
    "tags": []
}
