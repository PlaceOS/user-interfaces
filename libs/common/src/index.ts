export * from './lib/common';
export * from './lib/uploads';
export * from './lib/feature-available.guard';
export * from './lib/remote-logging.service';
export * from './lib/timezone-helpers';
export * from './lib/user-idle-time.service';
export * from './lib/booking-rules';
export * from './lib/keep-alive.service';
export * from './lib/uploads.service';
export * from './lib/shorten.fn';
