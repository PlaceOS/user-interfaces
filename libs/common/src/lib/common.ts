export { BaseClass } from './base.class';
export { GoogleAnalyticsService } from './google-analytics.service';
export { HotkeysService } from './hotkeys.service';
export { SettingsService } from './settings.service';
export { VorlonService } from './vorlon.service';
export { VERSION } from './version';

export * from './angular-animations';
export * from './api';
export * from './application';
export * from './formatting';
export * from './general';
export * from './notifications';
export * from './placeos';
export * from './room-configuration.interface';
export * from './settings';
export * from './types';
export * from './user-state';
export * from './spec-helpers';
