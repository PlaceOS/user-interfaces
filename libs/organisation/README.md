# organisation

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test organisation` to execute the unit tests.
