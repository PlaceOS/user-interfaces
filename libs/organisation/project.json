{
    "name": "organisation",
    "$schema": "../../node_modules/nx/schemas/project-schema.json",
    "projectType": "library",
    "sourceRoot": "libs/organisation/src",
    "prefix": "placeos",
    "targets": {
        "lint": {
            "executor": "@nx/eslint:lint",
            "outputs": ["{options.outputFile}"]
        },
        "test": {
            "executor": "@nx/jest:jest",
            "outputs": ["{workspaceRoot}/coverage/libs/organisation"],
            "options": {
                "jestConfig": "libs/organisation/jest.config.ts"
            }
        }
    },
    "tags": []
}
