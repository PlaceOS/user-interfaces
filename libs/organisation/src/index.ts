export * from './lib/building.class';
export * from './lib/building.utilities';
export * from './lib/level.class';
export * from './lib/organisation.class';
export * from './lib/organisation.service';
export * from './lib/organisation.utilities';
export * from './lib/desk.class';
export * from './lib/region.class';
