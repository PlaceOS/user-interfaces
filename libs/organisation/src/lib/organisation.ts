export * from './building.class';
export * from './building.utilities';
export * from './level.class';
export * from './organisation.class';
export * from './organisation.service';
export * from './organisation.utilities';
export * from './desk.class';
