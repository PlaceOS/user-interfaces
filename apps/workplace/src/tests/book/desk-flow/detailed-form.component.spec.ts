import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { MatFormFieldModule } from '@angular/material/form-field';
import { MatInputModule } from '@angular/material/input';
import { MatSelectModule } from '@angular/material/select';
import { createComponentFactory, Spectator } from '@ngneat/spectator/jest';
import { MockComponent } from 'ng-mocks';
import { of } from 'rxjs';
import {
    Booking,
    BookingFormService,
    generateBookingForm,
} from '@placeos/bookings';
import { SettingsService } from '@placeos/common';
import {
    DateFieldComponent,
    DurationFieldComponent,
    TimeFieldComponent,
    UserListFieldComponent,
} from '@placeos/form-fields';
import { OrganisationService } from '@placeos/organisation';
import { DeskFlowDetailedFormComponent } from 'apps/workplace/src/app/book/desk-flow/detailed-form.component';

describe('DetailBookSpaceFormComponent', () => {
    let spectator: Spectator<DeskFlowDetailedFormComponent>;
    const createComponent = createComponentFactory({
        component: DeskFlowDetailedFormComponent,
        providers: [
            {
                provide: BookingFormService,
                useValue: {
                    setOptions: jest.fn(),
                    options: of({}),
                    features: of(['one', 'two']),
                },
            },
            {
                provide: OrganisationService,
                useValue: {
                    building_list: of([{ id: 'bld-1' }, { id: 'bld-2' }]),
                    active_levels: of([{ id: 'lvl-1' }, { id: 'lvl-2' }]),
                },
            },
            { provide: SettingsService, useValue: { get: jest.fn() } },
        ],
        declarations: [
            MockComponent(DateFieldComponent),
            MockComponent(TimeFieldComponent),
            MockComponent(DurationFieldComponent),
            MockComponent(UserListFieldComponent),
        ],
        imports: [
            ReactiveFormsModule,
            FormsModule,
            MatFormFieldModule,
            MatInputModule,
            MatSelectModule,
        ],
    });

    beforeEach(() => (spectator = createComponent()));

    it('should create component', () => {
        expect(spectator.component).toBeTruthy();
    });

    it('should match snapshot', () => {
        expect(spectator.element).toMatchSnapshot();
        spectator.setInput({
            form: generateBookingForm(new Booking({ date: 1 })),
        });
        spectator.detectChanges();
        expect(spectator.element).toMatchSnapshot();
    });

    it.todo('should allow for recurrence');

    it.todo('should allow for group bookings');
});
