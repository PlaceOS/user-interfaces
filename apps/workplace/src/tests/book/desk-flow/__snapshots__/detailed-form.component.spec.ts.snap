// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailBookSpaceFormComponent should match snapshot 1`] = `
<div
  id="root1"
>
  
</div>
`;

exports[`DetailBookSpaceFormComponent should match snapshot 2`] = `
<div
  id="root1"
>
  <form
    class="ng-untouched ng-pristine ng-valid"
    novalidate=""
  >
    
    <div
      class="flex flex-col sm:flex-row space-x-0 sm:space-x-2 w-[640px] max-w-[calc(100%-2rem)] mx-auto"
    >
      <div
        class="flex flex-col flex-1 w-full sm:w-1/4"
      >
        <label>
          Date
        </label>
        <a-date-field
          class="ng-untouched ng-pristine ng-valid"
          formcontrolname="date"
        >
           Date and time must be in the future 
        </a-date-field>
      </div>
      <div
        class="flex flex-col flex-1 w-full sm:w-1/4"
      >
        <label>
          Building
        </label>
        <mat-form-field
          appearance="outline"
          class="mat-form-field mat-primary mat-form-field-type-mat-select mat-form-field-appearance-outline mat-form-field-can-float ng-untouched ng-pristine ng-valid _mat-animation-noopable"
        >
          <div
            class="mat-form-field-wrapper"
          >
            <div
              class="mat-form-field-flex"
            >
              <div
                class="mat-form-field-outline"
              >
                <div
                  class="mat-form-field-outline-start"
                />
                <div
                  class="mat-form-field-outline-gap"
                />
                <div
                  class="mat-form-field-outline-end"
                />
              </div>
              <div
                class="mat-form-field-outline mat-form-field-outline-thick"
              >
                <div
                  class="mat-form-field-outline-start"
                />
                <div
                  class="mat-form-field-outline-gap"
                />
                <div
                  class="mat-form-field-outline-end"
                />
              </div>
              
              
              
              <div
                class="mat-form-field-infix"
              >
                <mat-select
                  aria-autocomplete="none"
                  aria-disabled="false"
                  aria-expanded="false"
                  aria-haspopup="true"
                  aria-invalid="false"
                  aria-required="false"
                  class="mat-select mat-select-empty ng-untouched ng-pristine ng-valid"
                  placeholder="Select building"
                  role="combobox"
                  tabindex="0"
                >
                  <div
                    cdk-overlay-origin=""
                    class="mat-select-trigger"
                  >
                    <div
                      class="mat-select-value"
                    >
                      <span
                        class="mat-select-placeholder mat-select-min-line"
                      >
                        Select building
                      </span>
                      
                      
                    </div>
                    <div
                      class="mat-select-arrow-wrapper"
                    >
                      <div
                        class="mat-select-arrow"
                      />
                    </div>
                  </div>
                  
                </mat-select>
                <span
                  class="mat-form-field-label-wrapper"
                >
                  
                </span>
              </div>
              
            </div>
            
            <div
              class="mat-form-field-subscript-wrapper"
            >
              
              <div
                class="mat-form-field-hint-wrapper ng-trigger ng-trigger-transitionMessages"
                style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
              >
                
                <div
                  class="mat-form-field-hint-spacer"
                />
              </div>
              
            </div>
          </div>
        </mat-form-field>
      </div>
      
      <div
        class="flex flex-col flex-1 w-full sm:w-1/4"
      >
        <label>
          Level
        </label>
        <mat-form-field
          appearance="outline"
          class="mat-form-field mat-primary mat-form-field-type-mat-select mat-form-field-appearance-outline mat-form-field-can-float mat-form-field-disabled ng-untouched ng-pristine ng-valid _mat-animation-noopable"
        >
          <div
            class="mat-form-field-wrapper"
          >
            <div
              class="mat-form-field-flex"
            >
              <div
                class="mat-form-field-outline"
              >
                <div
                  class="mat-form-field-outline-start"
                />
                <div
                  class="mat-form-field-outline-gap"
                />
                <div
                  class="mat-form-field-outline-end"
                />
              </div>
              <div
                class="mat-form-field-outline mat-form-field-outline-thick"
              >
                <div
                  class="mat-form-field-outline-start"
                />
                <div
                  class="mat-form-field-outline-gap"
                />
                <div
                  class="mat-form-field-outline-end"
                />
              </div>
              
              
              
              <div
                class="mat-form-field-infix"
              >
                <mat-select
                  aria-autocomplete="none"
                  aria-disabled="true"
                  aria-expanded="false"
                  aria-haspopup="true"
                  aria-invalid="false"
                  aria-required="false"
                  class="mat-select mat-select-disabled mat-select-empty ng-untouched ng-pristine ng-valid"
                  ngmodel=""
                  placeholder="Any Level"
                  role="combobox"
                  tabindex="-1"
                >
                  <div
                    cdk-overlay-origin=""
                    class="mat-select-trigger"
                  >
                    <div
                      class="mat-select-value"
                    >
                      <span
                        class="mat-select-placeholder mat-select-min-line"
                      >
                        Any Level
                      </span>
                      
                      
                    </div>
                    <div
                      class="mat-select-arrow-wrapper"
                    >
                      <div
                        class="mat-select-arrow"
                      />
                    </div>
                  </div>
                  
                </mat-select>
                <span
                  class="mat-form-field-label-wrapper"
                >
                  
                </span>
              </div>
              
            </div>
            
            <div
              class="mat-form-field-subscript-wrapper"
            >
              
              <div
                class="mat-form-field-hint-wrapper ng-trigger ng-trigger-transitionMessages"
                style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
              >
                
                <div
                  class="mat-form-field-hint-spacer"
                />
              </div>
              
            </div>
          </div>
        </mat-form-field>
      </div>
      
    </div>
    <div
      class="flex flex-col flex-1 w-[640px] max-w-[calc(100%-2rem)] mx-auto"
    >
      <label>
        Reason
      </label>
      <mat-form-field
        appearance="outline"
        class="mat-form-field w-full mat-primary mat-form-field-type-mat-input mat-form-field-appearance-outline mat-form-field-can-float ng-untouched ng-pristine ng-valid _mat-animation-noopable"
      >
        <div
          class="mat-form-field-wrapper"
        >
          <div
            class="mat-form-field-flex"
          >
            <div
              class="mat-form-field-outline"
            >
              <div
                class="mat-form-field-outline-start"
              />
              <div
                class="mat-form-field-outline-gap"
              />
              <div
                class="mat-form-field-outline-end"
              />
            </div>
            <div
              class="mat-form-field-outline mat-form-field-outline-thick"
            >
              <div
                class="mat-form-field-outline-start"
              />
              <div
                class="mat-form-field-outline-gap"
              />
              <div
                class="mat-form-field-outline-end"
              />
            </div>
            
            
            
            <div
              class="mat-form-field-infix"
            >
              <input
                aria-invalid="false"
                aria-required="false"
                class="mat-input-element mat-form-field-autofill-control ng-untouched ng-pristine ng-valid cdk-text-field-autofill-monitored"
                data-placeholder="Reason for booking..."
                formcontrolname="description"
                matinput=""
                placeholder="Reason for booking..."
              />
              <span
                class="mat-form-field-label-wrapper"
              >
                
              </span>
            </div>
            
          </div>
          
          <div
            class="mat-form-field-subscript-wrapper"
          >
            
            <div
              class="mat-form-field-hint-wrapper ng-trigger ng-trigger-transitionMessages"
              style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
            >
              
              <div
                class="mat-form-field-hint-spacer"
              />
            </div>
            
          </div>
        </div>
      </mat-form-field>
    </div>
    
    
    <div
      class="flex flex-col flex-1 w-[640px] max-w-[calc(100%-2rem)] mx-auto"
    >
      <label>
        Desk Features
      </label>
      <mat-form-field
        appearance="outline"
        class="mat-form-field w-full mat-primary mat-form-field-type-mat-select mat-form-field-appearance-outline mat-form-field-can-float ng-untouched ng-pristine ng-valid _mat-animation-noopable"
      >
        <div
          class="mat-form-field-wrapper"
        >
          <div
            class="mat-form-field-flex"
          >
            <div
              class="mat-form-field-outline"
            >
              <div
                class="mat-form-field-outline-start"
              />
              <div
                class="mat-form-field-outline-gap"
              />
              <div
                class="mat-form-field-outline-end"
              />
            </div>
            <div
              class="mat-form-field-outline mat-form-field-outline-thick"
            >
              <div
                class="mat-form-field-outline-start"
              />
              <div
                class="mat-form-field-outline-gap"
              />
              <div
                class="mat-form-field-outline-end"
              />
            </div>
            
            
            
            <div
              class="mat-form-field-infix"
            >
              <mat-select
                aria-autocomplete="none"
                aria-disabled="false"
                aria-expanded="false"
                aria-haspopup="true"
                aria-invalid="false"
                aria-required="false"
                class="mat-select mat-select-empty mat-select-multiple ng-untouched ng-pristine ng-valid"
                multiple=""
                placeholder="Any Feature"
                role="combobox"
                tabindex="0"
              >
                <div
                  cdk-overlay-origin=""
                  class="mat-select-trigger"
                >
                  <div
                    class="mat-select-value"
                  >
                    <span
                      class="mat-select-placeholder mat-select-min-line"
                    >
                      Any Feature
                    </span>
                    
                    
                  </div>
                  <div
                    class="mat-select-arrow-wrapper"
                  >
                    <div
                      class="mat-select-arrow"
                    />
                  </div>
                </div>
                
              </mat-select>
              <span
                class="mat-form-field-label-wrapper"
              >
                
              </span>
            </div>
            
          </div>
          
          <div
            class="mat-form-field-subscript-wrapper"
          >
            
            <div
              class="mat-form-field-hint-wrapper ng-trigger ng-trigger-transitionMessages"
              style="opacity:1;transform:translateY(0%);0:opacity;1:transform;"
            >
              
              <div
                class="mat-form-field-hint-spacer"
              />
            </div>
            
          </div>
        </div>
      </mat-form-field>
    </div>
    
  </form>
  
</div>
`;
